### PR TITLE
splitDiagrams

### PR DESCRIPTION
**Pull request by sidharthv96**
## :bookmark_tabs: Summary

Splits all diagrams into lazy loaded modules.

## :straight_ruler: Design Decisions

Mermaid will only load whichever diagram is being used.

### :clipboard: Tasks

Make sure you

- [ ] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [ ] :computer: have added unit/e2e tests (if appropriate)
- [ ] :notebook: have added documentation (if appropriate)
- [ ] :bookmark: targeted `develop` branch

### CODE REVIEW

File: packages/mermaid/src/diagram-api/detectType.ts [hunk 1]
```diff
@@ -54,6 +55,39 @@ export const registerLazyLoadedDiagrams = (...diagrams: ExternalDiagramDefinitio
...
+        try {
+          getDiagram(key);
+        } catch (error) {
+          try {
+            // Register diagram if it is not already registered
+            const { diagram, id } = await loader();
+            registerDiagram(id, diagram, detector);
+          } catch (err) {
+            // Remove failed diagram from detectors
+            log.error(`Failed to load external diagram with key ${key}. Remozing from detectors.`);
```
Comment (pbrolin47) [inline packages/mermaid/src/diagram-api/detectType.ts:73]: Minor typo. Change 'Remozing from detectors' to 'Removing from detectors'

File: packages/mermaid/src/diagrams/c4/c4Detector.ts [hunk 2]
```diff
@@ -1,5 +1,20 @@
-import type { DiagramDetector } from '../../diagram-api/types';
+import type { ExternalDiagramDefinition } from '../../diagram-api/types';
 
-export const c4Detector: DiagramDetector = (txt) => {
+const id = 'c4';
+
+const detector = (txt: string) => {
```
Comment (pbrolin47) [inline packages/mermaid/src/diagrams/c4/c4Detector.ts:5]: It seems the majority of other diagram detectors have a declaration "const detector: DiagramDetector = ...",
but in this file the detector type is not present. It's only declared as "const detector = ..."

File: packages/mermaid/src/diagrams/er/erDiagram.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// @ts-ignore: TODO Fix ts errors
+import erParser from './parser/erDiagram';
+import erDb from './erDb';
+import erRenderer from './erRenderer';
+import erStyles from './styles';
+
+export const diagram = {
```
Comment (pbrolin47) [inline packages/mermaid/src/diagrams/er/erDiagram.ts:7]: Maybe add type definition "DiagramDefinition" as "export const diagram: DiagramDefinition = {"

File: packages/mermaid/src/diagrams/flowchart/flowDiagram-v2.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+// @ts-ignore: TODO Fix ts errors
+import flowParser from './parser/flow';
+import flowDb from './flowDb';
+import flowRendererV2 from './flowRenderer-v2';
+import flowStyles from './styles';
+import { MermaidConfig } from '../../config.type';
+import { setConfig } from '../../config';
+
+export const diagram = {
```
Comment (pbrolin47) [inline packages/mermaid/src/diagrams/flowchart/flowDiagram-v2.ts:9]: Maybe add type definition "DiagramDefinition" as "export const diagram: DiagramDefinition = {"

File: packages/mermaid/src/diagrams/flowchart/flowDiagram.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+// @ts-ignore: TODO Fix ts errors
+import flowParser from './parser/flow';
+import flowDb from './flowDb';
+import flowRenderer from './flowRenderer';
+import flowRendererV2 from './flowRenderer-v2';
+import flowStyles from './styles';
+import { MermaidConfig } from '../../config.type';
+
+export const diagram = {
```
Comment (pbrolin47) [inline packages/mermaid/src/diagrams/flowchart/flowDiagram.ts:9]: Maybe add type definition "DiagramDefinition" as "export const diagram: DiagramDefinition = {"

**Review by pbrolin47 — COMMENTED** [review]
Reviewed.

**Review by aloisklink — COMMENTED** [review]
Looks good to me!

Two quick questions:
  - Are we still publishing a giant `mermaid.esm.min.mjs` bundle that has everything in a single-file (i.e. without any lazy-loading)?
    I know there's still quite a few people that just copy a single file `mermaid.min.js` file to their projects (e.g. this is what joplin does, see https://github.com/laurent22/joplin/tree/dev/packages/renderer/assets/mermaid).
    (The `mermaid-cli` project used to do the same thing, but we recently had to switch to using `vite` to get mermaid-mindmaps working).
  - We should probably document this change as a **BREAKING CHANGE** in our release notes too. I think pretty much every bundler should support `lazyLoading`, but just in case, we should document it.
    Plus, this might also affect people using websites in offline mode, e.g. in [PWA](https://en.wikipedia.org/wiki/Progressive_web_app), as lazy-loading probably needs internet connection.

**Pull request merged**